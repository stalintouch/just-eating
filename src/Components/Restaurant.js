import React, { Component } from 'react';
import fetchJSON from '../scripts/fetchJSON';
import RestaurantCard from './RestaurantCard';
import NotFound from './NotFound';
import CircularProgress from 'material-ui/CircularProgress';
import FlatButton from 'material-ui/FlatButton';
const imagesUrl =
  'https://api.unsplash.com/search/photos?client_id=cdaa537bca08095f8342feb8bc2e12893de05e62eba059daed8dba9e13e046bd&page=1&query=food&per_page=100';

class Restaurant extends Component {
  constructor(props) {
    super(props);
    this.state = { backgroundImages: [], loading: true };
    this.handleFocus = this.handleFocus.bind(this);
  }
  // fetches the background images from unsplash, to display them as background
  async componentDidMount() {
    const images = await fetchJSON(imagesUrl);
    console.log(images);
    this.setState({ backgroundImages: images.results });
  }

  componentWillReceiveProps() {
    // this settimeout is to make sure the loading spinner disappears after the component receive
    // props but not too soon so it shows the not found component
    setTimeout(() => {
      this.setState({ loading: false });
    }, 1000);
  }

  handleFocus() {
    this.props.focus('autoComplete');
  }
  // renderCards loops throught the props received from App.js and assembles a RestaurantCard with the
  // proper info passed
  renderCards = () => {
    return this.props.restaurants.map((restaurant, idx) => {
      return (
        <RestaurantCard
          backgroundImage={this.state.backgroundImages[idx].urls.full}
          key={restaurant.id}
          restaurantInfo={restaurant}
        />
      );
    });
  };

  render() {
    // shows the loading spinner
    if (!this.props.restaurants.length && this.state.loading) {
      return (
        <div style={{ display: ' flex', justifyContent: 'center' }}>
          <CircularProgress size={180} thickness={15} />
        </div>
      );
      // shows the not found component
    } else if (!this.props.restaurants.length && !this.state.loading) {
      return <NotFound />;
    } else {
      return (
        <div>
          <div style={{ marginBottom: '30px' }}>
            Not in {this.props.city}? Change it&nbsp;&nbsp;
            <FlatButton
              className="restaurant"
              label="Here"
              onClick={this.handleFocus}
              backgroundColor="#FFFF00"
            />
          </div>
          {/* shows the RestaurantCard component */}
          {this.renderCards()}
        </div>
      );
    }
  }
}

export default Restaurant;
