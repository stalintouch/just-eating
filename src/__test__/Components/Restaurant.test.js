import Restaurant from '../../Components/Restaurant';
import imagesData from './data/imagesData.json';
import restaurantData from './data/restaurantData.json';
import CircularProgress from 'material-ui/CircularProgress';

describe('<Restaurant />', () => {
  let wrapper;
  let wrapperMount;
  const restaurants = {
    restaurants: restaurantData
  };
  const focus = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<Restaurant {...restaurants} focus={focus} />);
    wrapper.setState({ backgroundImages: imagesData });
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('render a spinner when the component initial renders and no restaurant array has been loaded to state', () => {
    expect(wrapper.find(CircularProgress)).toHaveLength(1);
  });

  it('when clicked this button focus on the search bar, it needs to be delay until the component gets props and setstate to render', async () => {
    const button = await wrapper.find('#restaurant').first();

    await new Promise(resolve => {
      setTimeout(() => {
        resolve();
        button.simulate('click', { preventDefault() {} });
        expect(focus).toBeCalled();
      }, 1000);
    });
  });
});
