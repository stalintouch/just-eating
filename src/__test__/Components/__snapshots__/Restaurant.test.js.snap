// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Restaurant /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Restaurant
    focus={[Function]}
    restaurants={
        Object {
            "current_page": 1,
            "per_page": 25,
            "restaurants": Array [
              Object {
                "address": "1 Benvenuto Place",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 21307,
                "image_url": "https://www.opentable.com/img/restimages/21307.jpg",
                "lat": 43.68207,
                "lng": -79.40041,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=21307",
                "name": "Scaramouche Restaurant",
                "phone": "4169618011",
                "postal_code": "M4V 2L1",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=21307",
                "state": "ON",
              },
              Object {
                "address": "49 Front St E",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 82957,
                "image_url": "https://www.opentable.com/img/restimages/82957.jpg",
                "lat": 43.648033,
                "lng": -79.374377,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=82957",
                "name": "The Sultan's Tent",
                "phone": "4169610601x",
                "postal_code": "M5E 1B3",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=82957",
                "state": "ON",
              },
              Object {
                "address": "225 Ossington Ave",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 108505,
                "image_url": "https://www.opentable.com/img/restimages/108505.jpg",
                "lat": 43.649015,
                "lng": -79.420648,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=108505",
                "name": "Salt Wine Bar",
                "phone": "4165337258",
                "postal_code": "M6J 2Z8",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=108505",
                "state": "ON",
              },
              Object {
                "address": "2537 Yonge St.",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 3637,
                "image_url": "https://www.opentable.com/img/restimages/3637.jpg",
                "lat": 43.713,
                "lng": -79.3996,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=3637",
                "name": "North 44",
                "phone": "4164874897x1",
                "postal_code": "M4P 2H9",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=3637",
                "state": "ON",
              },
              Object {
                "address": "45 Colborne Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 105922,
                "image_url": "https://www.opentable.com/img/restimages/105922.jpg",
                "lat": 43.648996,
                "lng": -79.375032,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=105922",
                "name": "Woods Restaurant & Bar",
                "phone": "4162149918x",
                "postal_code": "M5E 1E3",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=105922",
                "state": "ON",
              },
              Object {
                "address": "169 Niagara Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 91948,
                "image_url": "https://www.opentable.com/img/restimages/91948.jpg",
                "lat": 43.641939,
                "lng": -79.406543,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=91948",
                "name": "Edulis Restaurant",
                "phone": "4167034222",
                "postal_code": "M5V 1C9",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=91948",
                "state": "ON",
              },
              Object {
                "address": "655 Dixon Rd.",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 44791,
                "image_url": "https://www.opentable.com/img/restimages/44791.jpg",
                "lat": 43.6898,
                "lng": -79.578,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=44791",
                "name": "Ginko Japanese Restaurant",
                "phone": "4162488445",
                "postal_code": "M5H 3M7",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=44791",
                "state": "ON",
              },
              Object {
                "address": "1585 Dundas West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 75154,
                "image_url": "https://www.opentable.com/img/restimages/75154.jpg",
                "lat": 43.649767,
                "lng": -79.434577,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=75154",
                "name": "Lula Lounge",
                "phone": "4165880307",
                "postal_code": "M6K 1T5",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=75154",
                "state": "ON",
              },
              Object {
                "address": "214 King St. West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 14806,
                "image_url": "https://www.opentable.com/img/restimages/14806.jpg",
                "lat": 43.647429,
                "lng": -79.386774,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=14806",
                "name": "Il Fornello on King",
                "phone": "4169772855",
                "postal_code": "M5H 3S6",
                "price": 3,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=14806",
                "state": "ON",
              },
              Object {
                "address": "190 University Ave",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 150661,
                "image_url": "https://www.opentable.com/img/restimages/150661.jpg",
                "lat": 43.649368,
                "lng": -79.385921,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=150661",
                "name": "Momofuku DAISHÅŒ - Toronto",
                "phone": "6472538000",
                "postal_code": "M5H 0A3",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=150661",
                "state": "ON",
              },
              Object {
                "address": "301 Front St West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 40507,
                "image_url": "https://www.opentable.com/img/restimages/40507.jpg",
                "lat": 43.644011,
                "lng": -79.388449,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=40507",
                "name": "360 The Restaurant at the CN Tower",
                "phone": "4163625411x",
                "postal_code": "M5V 2T6",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=40507",
                "state": "ON",
              },
              Object {
                "address": "220 King Street West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 112033,
                "image_url": "https://www.opentable.com/img/restimages/112033.jpg",
                "lat": 43.647258,
                "lng": -79.38695,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=112033",
                "name": "El Caballito Tequila Bar",
                "phone": "4166289838x",
                "postal_code": "M5H 1K4",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=112033",
                "state": "ON",
              },
              Object {
                "address": "79 Wellington St. West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 114148,
                "image_url": "https://www.opentable.com/img/restimages/114148.jpg",
                "lat": 43.646989,
                "lng": -79.381283,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=114148",
                "name": "Stratus",
                "phone": "4168651924x",
                "postal_code": "M5K 1J5",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=114148",
                "state": "ON",
              },
              Object {
                "address": "591 College Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 108499,
                "image_url": "https://www.opentable.com/img/restimages/108499.jpg",
                "lat": 43.655204,
                "lng": -79.413547,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=108499",
                "name": "Trattoria Taverniti",
                "phone": "4165370078x",
                "postal_code": "M6G 1B2",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=108499",
                "state": "ON",
              },
              Object {
                "address": "54 The Explanade,",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 70072,
                "image_url": "https://www.opentable.com/img/restimages/70072.jpg",
                "lat": 43.646911,
                "lng": -79.374339,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=70072",
                "name": "The Old Spaghetti Factory - Toronto",
                "phone": "4168649761",
                "postal_code": "M5E 1A6",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=70072",
                "state": "ON",
              },
              Object {
                "address": "25 Dockside Drive",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 61609,
                "image_url": "https://www.opentable.com/img/restimages/61609.jpg",
                "lat": 43.643513,
                "lng": -79.366927,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=61609",
                "name": "Against the Grain Urban Tavern-Corus Quay",
                "phone": "6473441562",
                "postal_code": "M5A 0B5",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=61609",
                "state": "ON",
              },
              Object {
                "address": "638 Queen Street West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 118675,
                "image_url": "https://www.opentable.com/img/restimages/118675.jpg",
                "lat": 43.646076,
                "lng": -79.409564,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=118675",
                "name": "Lisa Marie",
                "phone": "6477486822x",
                "postal_code": "M6J 1E3",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=118675",
                "state": "ON",
              },
              Object {
                "address": "111 Dupont St",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 115321,
                "image_url": "https://www.opentable.com/img/restimages/115321.jpg",
                "lat": 43.676026,
                "lng": -79.401205,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=115321",
                "name": "Playa Cabana",
                "phone": "4169293911",
                "postal_code": "M5R 1V4",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=115321",
                "state": "ON",
              },
              Object {
                "address": "81 Church Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 89974,
                "image_url": "https://www.opentable.com/img/restimages/89974.jpg",
                "lat": 43.651765,
                "lng": -79.375018,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=89974",
                "name": "Kirei Sushi Bar",
                "phone": "4162149393",
                "postal_code": "M5C2G2",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=89974",
                "state": "ON",
              },
              Object {
                "address": "55 Eglinton Ave. East",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 63499,
                "image_url": "https://www.opentable.com/img/restimages/63499.jpg",
                "lat": 43.707061,
                "lng": -79.396159,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=63499",
                "name": "Fin Izakaya",
                "phone": "6473473864",
                "postal_code": "M4P 1G8",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=63499",
                "state": "ON",
              },
              Object {
                "address": "60 Harbour Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 17146,
                "image_url": "https://www.opentable.com/img/restimages/17146.jpg",
                "lat": 43.642422,
                "lng": -79.37642,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=17146",
                "name": "Harbour Sixty",
                "phone": "4167772111",
                "postal_code": "M5J 1B7",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=17146",
                "state": "ON",
              },
              Object {
                "address": "150 York Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 113926,
                "image_url": "https://www.opentable.com/img/restimages/113926.jpg",
                "lat": 43.649354,
                "lng": -79.384684,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=113926",
                "name": "Drake One Fifty",
                "phone": "4163636150x",
                "postal_code": "M5H 3S5",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=113926",
                "state": "ON",
              },
              Object {
                "address": "4150 Yonge Street",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 5970,
                "image_url": "https://www.opentable.com/img/restimages/5970.jpg",
                "lat": 43.746912,
                "lng": -79.407363,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=5970",
                "name": "Auberge du Pommier",
                "phone": "4162222220x",
                "postal_code": "M2P 2C6",
                "price": 4,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=5970",
                "state": "ON",
              },
              Object {
                "address": "1588 Dundas St W",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 115669,
                "image_url": "https://www.opentable.com/img/restimages/115669.jpg",
                "lat": 43.649959,
                "lng": -79.434732,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=115669",
                "name": "Cafe Bar Pasta Inc.",
                "phone": "4165344794",
                "postal_code": "M6K 1T8",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=115669",
                "state": "ON",
              },
              Object {
                "address": "21 St Clair West",
                "area": "Toronto / SW Ontario",
                "city": "Toronto",
                "country": "CA",
                "id": 113791,
                "image_url": "https://www.opentable.com/img/restimages/113791.jpg",
                "lat": 43.687826,
                "lng": -79.394971,
                "mobile_reserve_url": "http://mobile.opentable.com/opentable/?restId=113791",
                "name": "Union Social Eatery St. Clair",
                "phone": "4169013456x",
                "postal_code": "M4V 1K6",
                "price": 2,
                "reserve_url": "http://www.opentable.com/single.aspx?rid=113791",
                "state": "ON",
              },
            ],
            "total_entries": 321,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <CircularProgress
        max={100}
        min={0}
        mode="indeterminate"
        size={180}
        thickness={15}
        value={0}
/>,
      "style": Object {
        "display": " flex",
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "max": 100,
        "min": 0,
        "mode": "indeterminate",
        "size": 180,
        "thickness": 15,
        "value": 0,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <CircularProgress
          max={100}
          min={0}
          mode="indeterminate"
          size={180}
          thickness={15}
          value={0}
/>,
        "style": Object {
          "display": " flex",
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "max": 100,
          "min": 0,
          "mode": "indeterminate",
          "size": 180,
          "thickness": 15,
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
